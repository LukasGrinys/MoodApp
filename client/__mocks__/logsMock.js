export const testLogs = [
    {
        id: '001',
        rating: 10,
        date: "TEST-DATE",
        text: "Test text"
    },
    {
        id: '002',
        rating: 9,
        date: "TEST-DATE",
        text: "Test text"
    },
    {
        id: '003',
        rating: 8,
        date: "TEST-DATE",
        text: "Test text"
    },
    {
        id: '004',
        rating: 7,
        date: "TEST-DATE",
        text: "Test text"
    }
];

export const testLog = {
    id: '001',
    rating: 10,
    timing: 'Afternoon', 
    date: "TEST-DATE",
    text: "Test text"
}