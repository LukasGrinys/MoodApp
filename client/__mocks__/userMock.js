export const testUser = {
    id: '001',
    firstName: 'test',
    lastName: 'test',
    email: 'test@test.com'
}