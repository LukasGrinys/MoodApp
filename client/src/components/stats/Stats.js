import React, {Component} from 'react';
import { connect } from 'react-redux';
import { getAllLogs } from './../../actions';

// Widgets
import Loading from './../../widgets/loading';
import BackButton from './../../widgets/backButton';
import GraphCanvas from './GraphCanvas';
import DaytimeGraph from './DaytimeGraph';
import styles from './stats.module.css';
// Functions
import { countAverage, lastSevenDaysAverage, listToArray, returnDaytimeStats} from './statsFunctions';
import { ratingButtonStyle } from './../../widgets/nightmodeColors';
import { useTheme } from './../../contexts/ThemeContext';

const AverageBox = ( {children }) => {
    const darkTheme = useTheme();
    return (
        <div className={styles.average_box} style={ratingButtonStyle(darkTheme)}>
            {children}
        </div>
    )
}

class Stats extends Component {
    state = {
        dataReceived: false
    }
    UNSAFE_componentWillReceiveProps(nextprops){
        if (nextprops.user.data && this.state.dataReceived === false) {
            this.props.dispatch(getAllLogs(nextprops.user.data.id));
            this.setState({
                dataReceived: true
            })
        }
    }
    render() {
        if (this.props.user.data && this.props.logs.list) {
            if (this.props.logs.list.length > 0) {
                const sevenDayAverageObject = lastSevenDaysAverage(this.props.logs.list);
                const listToArrayObject = listToArray(this.props.logs.list);
                const daytimeAverageObject = returnDaytimeStats(this.props.logs.list);
                return (
                    <div className={styles.container}>
                        <BackButton/>
                        <h1>User stats</h1>
                        <div className={styles.first_block}>
                            <div className={styles.average_block}>
                                <b>Your average mood:</b>
                                <AverageBox>{countAverage(this.props.logs.list)}</AverageBox>
                                <div className={styles.small_text_box}>Retrieved from the last 100 logs</div>
                            </div>
                            <div className={styles.average_block}>
                                <b>7 day mood:</b>
                                <AverageBox>{sevenDayAverageObject.average}</AverageBox>
                                <div className={styles.small_text_box}>Counted from the last 7 active days from <br/>
                                {sevenDayAverageObject.firstDay} to {sevenDayAverageObject.lastDay}</div>
                            </div>
                        </div>
                        <div className={styles.graph_container}>
                            <b>Your mood graph</b>
                            <GraphCanvas list={listToArrayObject.arr}/>
                            <div> From {listToArrayObject.firstDay} to {listToArrayObject.lastDay}</div>
                        </div>
                        <div className={styles.daytime_container}>
                            <b>Your mood throughout the day:</b><br/>
                            <DaytimeGraph obj={daytimeAverageObject}/>
                        </div>
                    </div>
                )
            } else {
                return (
                    <div>
                        <BackButton/>
                        No logs to be found for the statistics
                    </div>
                )
            }
        } else {
            return <Loading/>
        }
        
    }
};

function mapStateToProps(state){
    return {
        user: state.user,
        logs: state.logs
    }
}
export default connect(mapStateToProps)(Stats);