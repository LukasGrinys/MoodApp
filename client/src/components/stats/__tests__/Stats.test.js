import React from 'react';
import Stats from '../Stats';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { testLogs } from '../../../../__mocks__/logsMock';

const mockStore = configureStore();
const mockDispatch = jest.fn();

jest.mock("react-redux", () => {
    const { Provider, useSelector } = jest.requireActual("react-redux");
  
    return {
        useDispatch: () => mockDispatch,
        useSelector,
        Provider
    };
});

const testId = 'testId';

describe('Stats: loading data', () => {
    const store = mockStore({
        user: {
            userData: {
                id: testId
            }
        },
        logs: {
            allLogs: {
                logs: [],
                isFetching: true,
                error: null
            }
        }
    });

    const wrapper = mount(
        <Provider store={store}>
            <Stats/>
        </Provider>
    );

    test('Renders properly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Has loading widget', () => {
        expect(wrapper.find('Loading').length).toBe(1);
    });

    test('Dispatches getLogs action', () => {
        const getLogs = jest.fn();

        setTimeout( () => {
            expect(getLogs).toHaveBeenCalledTimes(1);
            expect(getLogs).toHaveBeenCalledWith({
                userId : testId,
                limit: 100
            });
            done();
        }, 0);
    })
});

describe('Stats with logs fetched', () => {
    const store = mockStore({
        user: {
            userData: {
                id: testId
            }
        },
        logs: {
            allLogs: {
                logs: testLogs,
                isFetching: false,
                error: null
            }
        }
    });

    const wrapper = mount(
        <Provider store={store}>
            <Stats/>
        </Provider>
    );

    test('Renders properly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Renders two average boxes', () => {
        expect(wrapper.find('AverageBox').length).toBe(2);
    });

    test('Renders GraphCanvas', () => {
        expect(wrapper.find('GraphCanvas').length).toBe(1);
    });

    test('Renders Daytime graph', () => {
        expect(wrapper.find('DaytimeGraph').length).toBe(1);
    });
});

test('Renders error if there was error fetching logs for stats', () => {
    const testError = 'Test error';

    const store = mockStore({
        user: {
            userData: {
                id: testId
            }
        },
        logs: {
            allLogs: {
                logs: [],
                isFetching: false,
                error: testError
            }
        }
    });

    const wrapper = mount(
        <Provider store={store}>
            <Stats/>
        </Provider>
    );

    expect(wrapper.find('ErrorBox').length).toBe(1);
    expect(wrapper.find('ErrorBox').text()).toBe(testError);
})