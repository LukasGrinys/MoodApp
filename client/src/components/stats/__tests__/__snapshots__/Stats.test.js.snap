// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stats with logs fetched Renders properly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Stats>
    <BackButton>
      <div
        className="container"
      >
        <div
          className="button"
          onClick={[Function]}
        >
          <FontAwesome
            name="arrow-left"
          >
            <span
              aria-hidden={true}
              className="fa fa-arrow-left"
            />
          </FontAwesome>
          <div
            className="text"
          >
            Go back
          </div>
        </div>
      </div>
    </BackButton>
    <div
      className="container"
    >
      <h1>
        User stats
      </h1>
      <div
        className="first_block"
      >
        <div
          className="average_block"
        >
          <p>
            <strong>
              Your average mood:
            </strong>
          </p>
          <AverageBox>
            <div
              className="averageBox"
            >
              8.5
            </div>
          </AverageBox>
          <div
            className="small_text_box"
          >
            Retrieved from the last 100 logs
          </div>
        </div>
        <div
          className="average_block"
        >
          <b>
            7 day mood:
          </b>
          <AverageBox>
            <div
              className="averageBox"
            >
              8.5
            </div>
          </AverageBox>
          <div
            className="small_text_box"
          >
            Counted from the last 7 active days from 
            <br />
            TEST-DATE
             to 
            TEST-DATE
          </div>
        </div>
      </div>
      <GraphCanvas
        logs={
          Array [
            Object {
              "date": "TEST-DATE",
              "id": "001",
              "rating": 10,
              "text": "Test text",
            },
            Object {
              "date": "TEST-DATE",
              "id": "002",
              "rating": 9,
              "text": "Test text",
            },
            Object {
              "date": "TEST-DATE",
              "id": "003",
              "rating": 8,
              "text": "Test text",
            },
            Object {
              "date": "TEST-DATE",
              "id": "004",
              "rating": 7,
              "text": "Test text",
            },
          ]
        }
      >
        <div
          className="graph_container"
        >
          <p>
            <strong>
              Your mood graph
            </strong>
          </p>
          <div
            className="graph_canvas"
          >
            <div
              className="y_axis_number"
            >
              10
            </div>
            <div
              className="y_axis_number"
              style={
                Object {
                  "transform": "translate(10px, 100px)",
                }
              }
            >
              5
            </div>
            <div
              className="y_axis_number"
              style={
                Object {
                  "transform": "translate(10px, 200px)",
                }
              }
            >
              0
            </div>
            <svg
              height="100%"
              width="100%"
            >
              <polyline
                points="1,20 490,20"
                style={
                  Object {
                    "fill": "none",
                    "stroke": "#cccccc",
                    "strokeWidth": "1",
                  }
                }
              />
              <polyline
                points="1,40 490,40"
                style={
                  Object {
                    "fill": "none",
                    "stroke": "#cccccc",
                    "strokeWidth": "1",
                  }
                }
              />
              <polyline
                points="1,60 490,60"
                style={
                  Object {
                    "fill": "none",
                    "stroke": "#cccccc",
                    "strokeWidth": "1",
                  }
                }
              />
              <polyline
                points="1,80 490,80"
                style={
                  Object {
                    "fill": "none",
                    "stroke": "#cccccc",
                    "strokeWidth": "1",
                  }
                }
              />
              <polyline
                points="1,100 490,100"
                style={
                  Object {
                    "fill": "none",
                    "stroke": "#cccccc",
                    "strokeWidth": "1",
                  }
                }
              />
              <polyline
                points="1,120 490,120"
                style={
                  Object {
                    "fill": "none",
                    "stroke": "#cccccc",
                    "strokeWidth": "1",
                  }
                }
              />
              <polyline
                points="1,140 490,140"
                style={
                  Object {
                    "fill": "none",
                    "stroke": "#cccccc",
                    "strokeWidth": "1",
                  }
                }
              />
              <polyline
                points="1,160 490,160"
                style={
                  Object {
                    "fill": "none",
                    "stroke": "#cccccc",
                    "strokeWidth": "1",
                  }
                }
              />
              <polyline
                points="1,180 490,180"
                style={
                  Object {
                    "fill": "none",
                    "stroke": "#cccccc",
                    "strokeWidth": "1",
                  }
                }
              />
              <polyline
                points="1,200 490,200"
                style={
                  Object {
                    "fill": "none",
                    "stroke": "#cccccc",
                    "strokeWidth": "1",
                  }
                }
              />
              <ReturnGraph
                points="0,20 0,40 0,60 0,80 "
              >
                <polyline
                  className="graph_stroke"
                  points="0,20 0,40 0,60 0,80 "
                />
              </ReturnGraph>
            </svg>
          </div>
          <div>
            From 
            TEST-DATE
             to 
            TEST-DATE
          </div>
        </div>
      </GraphCanvas>
      <DaytimeGraph
        logs={
          Array [
            Object {
              "date": "TEST-DATE",
              "id": "001",
              "rating": 10,
              "text": "Test text",
            },
            Object {
              "date": "TEST-DATE",
              "id": "002",
              "rating": 9,
              "text": "Test text",
            },
            Object {
              "date": "TEST-DATE",
              "id": "003",
              "rating": 8,
              "text": "Test text",
            },
            Object {
              "date": "TEST-DATE",
              "id": "004",
              "rating": 7,
              "text": "Test text",
            },
          ]
        }
      >
        <div
          className="daytime_container"
        >
          <p>
            <strong>
              Your mood throughout the day:
            </strong>
          </p>
          <div
            className="daytime_canvas"
          >
            <svg
              height="100%"
              overflow="hidden"
              width="100%"
            >
              <polyline
                points="1,60 300,60"
                style={
                  Object {
                    "fill": "none",
                    "stroke": "#cccccc",
                    "strokeWidth": "1",
                  }
                }
              />
              <polyline
                points="1,120 300,120"
                style={
                  Object {
                    "fill": "none",
                    "stroke": "#cccccc",
                    "strokeWidth": "1",
                  }
                }
              />
              <polyline
                points="1,180 300,180"
                style={
                  Object {
                    "fill": "none",
                    "stroke": "#cccccc",
                    "strokeWidth": "1",
                  }
                }
              />
              <polyline
                points="1,240 300,240"
                style={
                  Object {
                    "fill": "none",
                    "stroke": "#cccccc",
                    "strokeWidth": "1",
                  }
                }
              />
              <polyline
                points="1,300 300,300"
                style={
                  Object {
                    "fill": "none",
                    "stroke": "#cccccc",
                    "strokeWidth": "1",
                  }
                }
              />
            </svg>
            <div
              className="y_axis_number_daytime"
              style={
                Object {
                  "top": "-12px",
                }
              }
            >
              10
            </div>
            <div
              className="y_axis_number_daytime"
              style={
                Object {
                  "left": "-15px",
                  "top": "48px",
                }
              }
            >
              8
            </div>
            <div
              className="y_axis_number_daytime"
              style={
                Object {
                  "left": "-15px",
                  "top": "108px",
                }
              }
            >
              6
            </div>
            <div
              className="y_axis_number_daytime"
              style={
                Object {
                  "left": "-15px",
                  "top": "168px",
                }
              }
            >
              4
            </div>
            <div
              className="y_axis_number_daytime"
              style={
                Object {
                  "left": "-15px",
                  "top": "228px",
                }
              }
            >
              2
            </div>
            <div
              className="daytime_bottom_line"
            >
              <div
                className="daytime_time"
              >
                7:00
              </div>
              <div
                className="daytime_time"
                style={
                  Object {
                    "left": "33%",
                  }
                }
              >
                13:00
              </div>
              <div
                className="daytime_time"
                style={
                  Object {
                    "left": "66%",
                  }
                }
              >
                20:00
              </div>
              <div
                className="daytime_time"
                style={
                  Object {
                    "left": "100%",
                  }
                }
              >
                00:00
              </div>
              <img
                alt="Morning"
                className=""
                src={Object {}}
                style={
                  Object {
                    "left": "13%",
                    "padding": "2px",
                  }
                }
              />
              <img
                alt="Afternoon"
                className=""
                src={Object {}}
                style={
                  Object {
                    "left": "48%",
                  }
                }
              />
              <img
                alt="Evenining"
                className=""
                src={Object {}}
                style={
                  Object {
                    "left": "82%",
                  }
                }
              />
            </div>
          </div>
        </div>
      </DaytimeGraph>
    </div>
  </Stats>
</Provider>
`;

exports[`Stats: loading data Renders properly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Stats>
    <Loading>
      <div
        className="loadingScreen"
      >
        <div>
          Loading
        </div>
        <br />
        <div
          className="barContainer"
        >
          <div
            className="bar"
          />
          <div
            className="bar delay1"
          />
          <div
            className="bar delay2"
          />
        </div>
      </div>
    </Loading>
  </Stats>
</Provider>
`;
